ivadomed/loader/loader.py

```diff
@@ -757,14 +756,20 @@ def __getitem__(self, index):
         # Update metadata_input with metadata_roi
         metadata_gt = imed_loader_utils.update_metadata(metadata_input, metadata_gt)
 
-        if self.task == "segmentation":
+
+
+
+        if self.task in ["segmentation", "pose_estimation"]:
             # Run transforms on images
             stack_gt, metadata_gt = self.transform(sample=seg_pair_slice["gt"],
                                                    metadata=metadata_gt,
                                                    data_type="gt")
             # Make sure stack_gt is binarized
             if stack_gt is not None and not self.soft_gt:
                 stack_gt = imed_postpro.threshold_predictions(stack_gt, thr=0.5).astype(np.uint8)
+            if self.task == "pose_estimation":
+                stack_gt = torch.moveaxis(torch.squeeze(stack_gt), -1, 0)
```

Comment:
In the previous `ivadomed `you were checking the input is either `2D `or `3D`, if the data is `2D `then the [two_dim_compatible](https://github.com/ivadomed/ivadomed/blob/ccd82d22d31ffa53a2d6455ce3a94b97564d624a/ivadomed/transforms.py#L54) checks for the compatibility with adding a new dimension. However during the transformation, I generate new channels for the gt mask to demonstrate the intervertebral disc location and the `2D `check adds extra dimension `(1, H, W, C)` and I solved with the above code.  Please note that our input gt is `2D `but during the transformation, It becomes multichannel (`3D`).

Reply:
Just to make sure I'm understanding correctly...

1. `(H, W) => (H, W, 1)`: Because `Resize` has `two_dim_compatible` decorator.
2. `(H, W, 1) => (H, W, C)` : Because `VertebralSplitting` creates one channel per `num_joint` in [`get_posedata`](https://github.com/ivadomed/ivadomed/blob/a502b37bd988ea0eae2fd5fa002666a2003a16bb/ivadomed/transforms.py#L354-L363).
3. `(H, W, C) => (1, H, W, C)`: Because of `NumpyToTensor` results in a new axis (`np.array` makes new axis [here](https://github.com/ivadomed/ivadomed/blob/a502b37bd988ea0eae2fd5fa002666a2003a16bb/ivadomed/transforms.py#L207)).
4. `(1, H, W, C) => (H, W, C)`: Because of `torch.squeeze(stack_gt)`.
5. `(H, W, C) => (C, H, W)`: Because of `torch.moveaxis(stack_gt, -1, 0)`.

What I'm confused about is why we need to do step 5 here... Couldn't we make sure the image has the correct axis order inside step 2, when we create the multichannel image in `get_posedata`?

Reply:
I handled the axis order in step 2, hence, step 5 is removed.

Reply:
The only thing I'm still curious about is the `torch.squeeze(stack_gt)` operation. 

It is necessary because of `NumpyToTensor` in Step 3. But, I imagine other tasks must use `NumpyToTensor` as well, so I'm wondering why it is only applied for `self.task == "pose_estimation"`.

Reply:
The point is that the `sample `is a `nested list,` thus, `np.array()` adds a dimension. While we don't need a new dimension for pose data (It already has channels). That's why we remove it on the loader section, it can be transferred to this part also.

Reply:
My apologies, I'm still a bit confused.

> While we don't need a new dimension for pose data (It already has channels).

My understanding was that the extra 4th dimension was not for channels, but for the number of images in a batch, e.g. the `N` in `(N, C, H, W)`. So, I was thinking it should be OK in theory (no need for `np.squeeze`).

Comparing to other tasks/models (for example, [sc-gm_t2star_exvivo](https://github.com/ivadomed/sc-gm_t2star_exvivo)), I see it also [uses `NumpyToTensor`](https://github.com/ivadomed/sc-gm_t2star_exvivo/blob/master/sc-gm_t2star_exvivo.json#L136). But, there is no `np.squeeze` for `task=="segmentation"`, so to me it seems the extra dimension doesn't seem to make a difference.

Reply:
My motivation for asking these questions is: Apart from this `np.squeeze` operation, `task="pose_estimation"` is identical to `task="segmentation"` in `loader.py`.

Since it's so close to the existing task, I'm trying to make sure that it's even necessary to even create a new task type. If we can just group this new model under `task="segmentation"`, then it might make it a bit easier to maintain in the future.

Reply:
It depends on how you define the objective if the goal is to segment the intervertebral discs then yes it is a segmentation task but if the goal is to estimate the 3D position of each intervertebral disc then it poses estimation.  Inside the `loader.py`, three times I have checked the `pose_estimation`  task, where the first two were for the code clarity and the last one is required for a `squeeze`. We can remove the first two. For the last one since we are inside the 2D class `MRI2DSegmentationDataset` I can expect the gt to have 3 dimensions, thus, I can simply remove the extra dimension using shape condition. for example `if len(stack_gt.shape)>3: stack_gt = torch.squeeze(stack_g`t). What do you think?

Reply:
My thinking was that there shouldn't be any need for `torch.squeeze` at all... From `loader.py`'s perspective, it should be able to treat all `stack_gt` for  `task=="segmentation"` as the same form: `[N, <gt_img_dim>]`.

* In some cases, `<gt_img_dim>==[H, W]`, so stack_gt should be `[N, H, W]` (stack of regular 2D GT images).
* In your case, `<gt_img_dim>==[C, H, W]`, so stack_gt should be `[N, C, H, W]` (stack of layered 2D GT images).

But in all cases, we preserve the `N` dimension added by `NumpyToTensor`.

> I can expect the gt to have 3 dimensions

Why does GT need to have 3 dimensions? I was thinking that we should expect the GT to have 4 dimensions, because `VertebralSplitting` adds a dimension.

Reply:
> Why does GT need to have 3 dimensions?

Ah, I see... `MRI2DSegmentationDataset` specifically expects `[N, H, W]`. So, GT having a 4th `C` channel required a hack. 

It feels a bit unclear to force GT data [`C, H, W]` into the form `[N, H, W]`...

Reply:
>  For the last one since we are inside the 2D class `MRI2DSegmentationDataset` I can expect the gt to have 3 dimensions, thus, I can simply remove the extra dimension using shape condition. for example `if len(stack_gt.shape)>3: stack_gt = torch.squeeze(stack_g`t). What do you think?

It's OK, but it still sort of feels like the wrong place for this solution. I wish we could solve this in `VertebralSplitting`, since that is where the problem is introduced...

Reply:
Oh!! I have an idea! What if we don't use `@multichannel_capable` for `VertebralSplitting`?

In multichannel_capable, it interprets a list of images as multiple channels, right? 

https://github.com/ivadomed/ivadomed/blob/a502b37bd988ea0eae2fd5fa002666a2003a16bb/ivadomed/transforms.py#L34-L36

But, this causes a problem in `VertebralSplitting`, because we go from `list([H,W]) => list([C,H,W])`. Since lists are interpreted as multichannel images, `list([C,H,W])` actually means `[C, C, H, W]`. _That's_ why we get an extra dimension.

What we actually want to do is to have `VertebralSplitting` go from `list([H,W]) => list([H,W], [H,W], [H,W]...)`. And, we can do this if we get rid of `@multichannel_capable`.

Reply:
About this part `VertebralSplitting → list([H,W], [H,W], ...)` do you mean to generate a mask for each joint separately and append them in one list? Once we remove the `@multichannel_capable` it does not recognize the `metadata` as a dictionary.

Reply:
> About this part `VertebralSplitting → list([H,W], [H,W], ...)` do you mean to generate a mask for each joint separately and append them in one list? 

Yes, exactly. Making `ys_ch` a list rather than a `np.zeros` array. Something roughly similar to:

```diff
            for i in range(1, num_labels+1):
                y_i = msk * np.where(labels_im == i, 1, 0)
-               ys_ch[i-1, :,:] = y_i
+               ys_ch.append(y_i)
```

> Once we remove the `@multichannel_capable` it does not recognize the `metadata` as a dictionary. 

I'm not sure I understand what you mean... isn't `metadata` the same as before, just wrapped in a list?

Reply:
In a simple example just to check it is working or not:  we have data_out with shape `[C, H, W]`, so making list as your suggestion does not solve the problem. 
data_out2 = []
for idx in range(data_out.shape[0]):
    data_out2.append(data_out[idx])    
return data_out2, metadata
Let me know if I got your idea correctly. Thanks

Reply:
I've made a commit on a separate branch just to demonstrate the changes: https://github.com/ivadomed/ivadomed/commit/dae06a5516805964fa77b7e08a67b15082aa05d1

When I run your `config_pose.json` with `data-multi-subject`, I see that `stack_gt` has the desired shape:

![Screenshot from 2021-05-31 08-20-37](https://user-images.githubusercontent.com/16181459/120192528-45f29a80-c1e9-11eb-98e5-9a3d5003ef35.png)

Reply:
Great, it works well. Thanks.

Reply:
I think after this change, we might be able to get rid of all the changes from `loader.py` and `utils.py`, which could make the contribution a little lighter. :smile:

Reply:
Task type pose is completely removed and switched to segmentation.

Reply:
Thank you! Just a nitpick, but I think that there are still some whitespace/new line changes. (Check ["Files"](https://github.com/ivadomed/ivadomed/pull/787/files) view on this PR.)

Reply:
Thanks, @joshuacwnewton I will double-check it.

Reply:
Hi @rezazad68. The whitespace changes I was talking about above were not actually addressed.

In the [Files view](https://github.com/ivadomed/ivadomed/pull/787/files), you can see that the file `loader.py` still has whitespace changes.

![Screenshot from 2021-06-25 12-55-15](https://user-images.githubusercontent.com/16181459/123460051-0edb9300-d5b5-11eb-8ecd-dfc2f6dcf86a.png)

Please fix these before marking the conversation as resolved. Thanks! :slightly_smiling_face:

Reply:
Just to explain why I am requesting this, there are a few reasons why we do not want whitespace changes like these:

* They add clutter to the "Files" view, which makes it a little harder to review (more scrolling, harder to see what the "important" changes are).
* If this pull request is merged, then this file's ["History"](https://github.com/ivadomed/ivadomed/commits/master/ivadomed/loader/loader.py) will be polluted. So, it will look like that file was changed, even when it wasn't _really_ changed. (This makes it harder to go back and understand what was done.)

In general, it is helpful to try and minimize the "noise" in the changes you make, so that you only make meaningful changes.

Also, here is a [pretty good article](https://medium.com/google-developer-experts/how-to-pull-request-d75ac81449a5) about "pull request etiquette" (see point number 3).

Reply:
@joshuacwnewton thanks for the review, it is fixed. Thanks for the nice explanation :)

Reply:
Sorry to bother, but it still does not look fixed to me. :sweat_smile: 

![Screenshot from 2021-06-25 15-47-23](https://user-images.githubusercontent.com/16181459/123477895-b7492180-d5cc-11eb-951b-f97feffad4a2.png)

Reply:
It is actually fine, one line removed and then added. If you check the [previous ivadomed](https://github.com/ivadomed/ivadomed/blob/e78316337d40fe24adef8f18d3714bdeed99394f/ivadomed/loader/loader.py#L719 ) version we have an empty line there and because of that, I put it there.

Reply:
> It is actually fine, one line removed and then added. If you check the previous ivadomed version we have an empty line there and because of that

Ah, well... The file viewer shows that something has changed, so there is indeed a change. :sweat_smile: 

In this case, there are 2 removed blank lines (in red). There is also a change in whitespace... it goes from having no spaces (`""`) to having 12 spaces (`"____________"`).

I'm sorry for being so nitpicky here. My comments may seem a bit silly, because it honestly doesn't _really_ matter. But, cleaning up your PRs is a good skill to practice, so that you can fix these things easily in the future and save your reviewers/teammates the trouble. :stuck_out_tongue:

Reply:
All are solved.